### PR TITLE
Set thread names for game threads

### PR DESCRIPTION
**Pull request by p0358**
Adds nice thread names that can be visible in crash dumps, non-attachable debuggers and generally in all places where old method of throwing exceptions to attached debugger on game start wouldn't work:
image

Can't believe it wasn't already there lol, makes debugging pain without.

Also the method to convert charsets is bad, but that's not my fault, seems Northstar doesn't have any util func for that apparently either, and this "method" was used in some other places in the codebase where it didn't belong. As thread names are generally static it should be fine in this place, but still.

Tested to work as you can see on the screenshot.

### CODE REVIEW

File: primedev/core/tier0.cpp [hunk 1]
```diff
@@ -18,11 +20,31 @@ void TryCreateGlobalMemAlloc()
 	g_pMemAllocSingleton = CreateGlobalMemAlloc(); // if it already exists, this returns the preexisting IMemAlloc instance
 }
 
+AUTOHOOK_PROCADDRESS(ThreadSetDebugName, tier0.dll, ThreadSetDebugName, void, __fastcall, (HANDLE threadHandle, const char* name))
+{
+	if (threadHandle == 0)
+		threadHandle = GetCurrentThread();
+
+	// need to grab it dynamically as this function was only introduced at some point in Windows 10
```
[the remark this answers is not in the document]
Comment (p0358) [reply primedev/core/tier0.cpp:28]: well that's why it grabs the function dynamically and checks with `if` whether it got the address: on older systems simply nothing is gonna happen, threads won't be named with this API

**Comment by GeckoEidechse** [comment]
Uh, where can I see the thread names? ^^"

**Comment by p0358** [comment]
Many tools, like x64dbg, Very Sleepy debugger, probably in various "process explorer" kind of programs too.

Also worth noting for future reference from RoyalBlue:
> if you want names for miles threads as well you need to hook a func there as well because miles only names threads when a debugger is present on thread start
> its mileswin64 + 0x3EBD0 a1 is thread handle a2 is the name

https://discord.com/channels/920776187884732556/950322078945538058/1206634307976699905

The above and possibly bink could also be added, but unfortunately for now I cannot be arsed to do that. They might be done in a separate PR though, so I leave that for future reference here. This PR alone makes the situation much nicer already by naming game threads...

**Comment by p0358** [comment]
Oh also one more thing, this doesn't set the name of MainThread as the hook is done too late. The thread that does the hooking is 99% the main thread, but that should be tested and then when the hook is performed it should set the name of its own thread to "MainThread" then.

**Comment by r3muxd** [comment]
the real fix for this is to use tier0 to make threads so your TLS gets set up correctly

**Comment by GeckoEidechse** [comment]
I was told the comment above is not really relevant to the PR...

**Comment by GeckoEidechse** [comment]
Fails to build after updating with `main`. Any ideas @p0358 ^^

**Comment by p0358** [comment]
I guess the CModule class was refactored in master

**Comment by GeckoEidechse** [comment]
Any chance you could update the PR so that we can merge :point_right::point_left:

**Comment by GeckoEidechse** [comment]
CI still failing btw @p0358

**Comment by p0358** [comment]
obraz

This works now too. There's also MilesMixer thread, but it seems it's not being created, thus it's not on the list? idk why but welp

**Comment by p0358** [comment]
Merge with rebase maybe, I rebased it to two distinct commits on top of main

**Comment by GeckoEidechse** [comment]
Will consider, thanks <3

**Review by GeckoEidechse — APPROVED** [review]
Verified that it still runs fine on Linux. This is more of a breakage check than a functionality check though tbh...

**Comment by ASpoonPlaysGames** [comment]
Would appreciate if someone could give this code review, since I have switched the hooks over to be manually hooked instead of using autohook.

**Review by ASpoonPlaysGames — APPROVED** [review]
Works on my machine:
image
image

**Review by ASpoonPlaysGames — APPROVED** [review]
Code looks good (although I am a bit biased since I converted the hooks to manually hooking) other than the stuff indicated by the comments. Fixing that would be beyond the scope of the PR though so it shouldn't block anything

**Comment by ASpoonPlaysGames** [comment]
@F1F7Y Any chance you could sanity check the stuff that I wrote? Would prefer to not be reviewing my own code in a PR if possible

**Comment by p0358** [comment]
Hehe I can't physically do an approving review on my own PR lol. But yeah it does seem to look okay...

**Pull request merged**